### ISSUE TITLE
MTP model hangs the GPU and allocates unboundedly after #3695 — no tokens emitted (Qwen3.6-35B-A3B-oQ6-mtp)

### ISSUE THREAD

**Issue by shironexus-gif** [jundot/omlx#3702]
**Describe the bug**

After updating to a main that includes #3695 (`feat(mtp): support multi-request Lightning MTP`, merged as 4d4750ee), requests against an `-mtp` model stop producing tokens. Generation hangs, the GPU is reset by the watchdog, and process memory grows without bound until the machine is close to OOM.

The GPU fault originates in the MTP reconcile path — note this one says **`Caused`**, so it is the culprit rather than a victim:

```
omlx.patches.mlx_lm_mtp.batch_generator - WARNING - MTP reconcile failed, falling back to plain drop:
  [METAL] Command buffer execution failed: Caused GPU Hang Error
  (00000003:kIOGPUCommandBufferCallbackErrorHang).
```

Every other command buffer in flight is then discarded as collateral of the GPU recovery:

```
omlx.scheduler - ERROR - Chunked prefill failed for a3f5ab93-...:
  [METAL] Command buffer execution failed: Discarded (victim of GPU error/recovery)
  (00000005:kIOGPUCommandBufferCallbackErrorInnocentVictim).

omlx.scheduler - ERROR - Chunked prefill failed for eb377ff7-...:
  ANE input pack command buffer failed; canceling this dispatch:
  Error Domain=MTLCommandBufferErrorDomain Code=1 "Discarded (victim of GPU error/recovery)"
  UserInfo={NSUnderlyingError=... Domain=IOGPUCommandQueueErrorDomain Code=5}
```

after which the MTP state cannot be recovered and subsequent streams fail:

```
omlx.server - ERROR - Error during chat streaming: Lightning MTP could not restore batch calibration state
omlx.server - ERROR - POST /v1/chat/completions → 500 (unhandled)
```

**To Reproduce**

1. Serve `Qwen3.6-35B-A3B-oQ6-mtp` on a main that includes 4d4750ee (#3695).
2. Send a single streaming chat completion:
   ```bash
   curl -sN http://127.0.0.1:8001/v1/chat/completions \
     -H 'Content-Type: application/json' \
     -d '{"model":"Qwen3.6-35B-A3B-oQ6-mtp","stream":true,"max_tokens":120,
          "messages":[{"role":"user","content":"Count from 1 to 40, one number per line."}]}'
   ```
3. Observe that the request emits **zero** content chunks. I aborted it at 151 s.
4. While it runs, watch resident memory: it climbed at roughly 10 GB/s and had to be killed before it exhausted 512 GB of unified memory.
5. See the errors above in `~/.omlx/logs/server.log`.

It never reaches decode — it appears to hang during prefill/priming, which matches the user-visible symptom of "model is loaded, but no tokens ever come out".

**Expected behavior**

Tokens stream normally. Failing that, the request should be rejected by the prefill memory guard rather than allocating until the system is near OOM.

**Screenshots**

N/A — log output included above and below instead.

**Desktop (please complete the following information):**
 - macOS Version: 26.6.2
 - oMLX Version: 0.7.0.dev2
 - Hardware: M3 Ultra, 512 GB unified memory
 - MLX: 0.32.2
 - Xcode 27.0 / Metal Toolchain 27A266a
 - Custom kernels built and verified on the fast native path (`bonsai`, `glm_moe_dsa`, `minimax_m3`, `qwen35_prefill` all load natively, ABI ok)

**Additional context**

Why #3695 is the suspect:

- The fault is raised from `omlx/patches/mlx_lm_mtp/batch_generator.py`, the file that PR rewrote most heavily (+601/−388).
- `Lightning MTP could not restore batch calibration state` corresponds to the batch calibration/parking logic the PR introduces.
- The stall happens before any token is emitted, and the PR also reworked `prompt_priming.py` (+382/−34).
- The affected model is an MTP model, and 4d4750ee is an ancestor of the tree where this occurs.

**Comment by jundot** [comment]
@shironexus-gif I tested current main (`bd841fc3`) with `Qwen3.6-35B-A3B-oQ4-mtp` on an M3 Ultra with 512 GB unified memory and the same major macOS version. I rebuilt the custom kernels from current main and verified native loading and ABI compatibility.

| Test | Chunked Prefill OFF | Chunked Prefill ON |
|---|---|---|
| Your counting prompt, one request | Completed | Completed |
| Your counting prompt, two concurrent requests | Completed | Completed |
| 9,929-token prompt, one request | Completed | Completed with three prefill chunks |

Every request generated 120 tokens with MTP active, without a GPU error or runaway memory growth. I haven't identified the cause yet; this used the oQ4 checkpoint with ANE acceleration disabled. I'll also test on an M5 Max and share the results.

If you installed oMLX in a virtual environment, could you remove all previously built custom-kernel binaries and rebuild/reinstall them against the MLX version in that environment, then retry?

**Comment by jundot** [comment]
@shironexus-gif I also downloaded `Jundot/Qwen3.6-35B-A3B-oQ6-mtp` and repeated the tests on main `bd841fc3` with the rebuilt native kernels: your counting prompt with one and two concurrent requests, plus a 9,929-token prompt.

All four combinations of ANE OFF/ON and Chunked Prefill OFF/ON completed successfully. All 16 requests streamed 120 tokens with Lightning MTP active, with no GPU errors or runaway memory growth. Peak process RSS was about 28.8 GiB. I verified actual batched MTP execution and ANE calls when enabled.

I still haven't reproduced the issue on my M3 Ultra 512 GB. My macOS version is 26.5 (25F71), versus your reported 26.6.2; MLX is 0.32.2.

**Comment by jundot** [comment]
@shironexus-gif I found a likely cause in the late-join calibration path. When a new request joins an active MTP batch, rebuilding the existing requests' caches used the speculative verification path, retaining unnecessary per-token rollback states. I reproduced a roughly three-minute stall and about 155 GiB of peak MLX allocation, though not the GPU hang itself.

I changed cache rebuilding to use ordinary prefill. The same test now takes about 8-9 seconds to rebuild, with peak MLX allocation around 34 GiB. I also verified that the rebuilt Qwen caches and next-token results match ordinary prefill.

The fix is on main and will be included in `0.7.0.dev3`. Could you retry with that build and let me know whether it resolves the hang?

**Comment by shironexus-gif** [comment]
Hey @jundot first of all, thank you so much for the tremendous work you do (once i find my wallet you will receive a big coffee!). Never would have thought that you immediately look into it.

I added the newest commit and repeated the test. It still happens but it seems like the blockage gets cleared quicker.
Here are some screenshots, the model suddenly stops generating tokens:

<img width="1632" height="321" alt="Image" src="https://github.com/user-attachments/assets/90ad882a-8a79-451e-bee7-5af99556d6aa" />

<img width="699" height="100" alt="Image" src="https://github.com/user-attachments/assets/e517f456-2871-46c8-bfcc-38c15f979dfe" />

I am also still trying to investigate this issue because i also updated my OpenCode version to 2.0.4 this morning and they changed some plugin functionality which interacts with omlx. Though on the other hand when i did some internal omlx performance benchmarks i also noticed that something felt off when it came to the speed of qwen3.6-35b-a3b. Seemed generally way slower.

If i find something i will let you know immediately.

**Comment by jundot** [comment]
@shironexus-gif Your quick follow-up and screenshots helped me identify another cause: when a third request joined, the existing requests' full contexts were still being reprocessed. I fixed that in `65c65e38`, so the transition now reuses their caches.

Catching this before the dev release was a big help. Thank you for testing so quickly. When you have time, could you try the current main and check whether the same issue still occurs?

**Comment by shironexus-gif** [comment]
@jundot just as i started writing "Its no longer happening!" ... it happened .... I've collected all of my subagent runs and let an Agent analyze the overall situation. I will attach his response, take it with a big grain of salt since i lack the knowledge to truly know if something in there can be of value.

After your last commit we now have a wave pattern:

<img width="697" height="118" alt="Image" src="https://github.com/user-attachments/assets/dd0cb605-ad80-42d1-b31e-c72a0fb1f619" />

---

### 1. Root Cause Analysis: What Happened in the Screenshots

The symptoms in your screenshots:
* **All 3 requests stalled at `"last token 55s ago"`** at the exact same moment.
* **Effective speeds collapsed to `1.7 tok/s` (Request 1) and `0.2 tok/s` (Request 2)**.
* **Wired Memory surged to 240.22 GB** (with `"Enforcer disabled"` / memory guard off in settings).

#### The Core Mechanism: Full-Prompt Re-Prefill on Cache Reconcile
In `omlx/patches/mlx_lm_mtp/batch_generator.py`, whenever an MTP state transitions to standard decoding (due to late-joining requests, batch calibration, or parking), oMLX invokes `_reconcile_mtp_to_standard()`. 

The implementation of `_reconcile_mtp_to_standard`:
```python
tok_arr = _ensure_uint32(mx.array(list(tokens))) # tokens[0] is the ENTIRE history!
for start in range(0, total, step):
    logits = gen_batch.model(tok_arr[None, start : start + step], cache=new_cache)
```
It **re-prefills the entire prompt + generated token history from token 0 to $N$** in 512-token chunks into a newly allocated cache.

#### Why Short Tests Masked the Problem
* In @jundot’s reproduction test, the prompt was ~40 tokens (counting prompt). Re-prefilling 40 tokens takes **~8 milliseconds**.
* In your OpenCode subagent session (`subagent-sessions-forge-onboarding-review.md`), subagents read large codebases, Makefiles, and configs. Prompt lengths were **70,000 to 150,000 tokens per subagent**.
* Re-prefilling a 100,000-token prompt in 512-token chunks requires **195 full model forward passes** per request. For 2–3 concurrent requests, that is over **300,000 tokens of re-prefill** on a 35B model.

#### Explaining Every Metric
1. **The 55s Stall**: Prefilling 200k–300k tokens on an M3 Ultra takes 1.5 to 3 minutes. During this entire time, the inference thread is blocked in Python inside the re-prefill loop. **Zero tokens can be emitted**, causing all active requests to freeze simultaneously at `"last token 55s ago"`.
2. **The 1.7 tok/s and 0.2 tok/s Collapse**: Token rate is calculated as $\text{tokens emitted} / \text{total wall-clock time}$. Because the engine spent 95% of its wall-clock time paused in prompt re-prefill loops:
   * Request 1: 132 tokens in 76s = 1.7 tok/s.
   * Request 2: 10 tokens in 57s = 0.17 tok/s (0.2 tok/s).
   The decode kernel itself was not slow; it was simply preempted by massive re-prefills.
3. **The 240 GB Wired Memory Surge**:
   * For every request reconciled, `_rebuild_singleton_cache` allocates an entirely new 100k-token KV cache alongside the existing cache.
   * `_merge_row_caches` then allocates a merged batch cache.
   * On Apple Silicon, Metal buffers in the MLX pool are held as macOS **Wired Memory**. Because `"prefill_memory_guard": false` was set in `~/.omlx/settings.json`, there was no ceiling to prevent memory from climbing to 240 GB.

---

### 2. Evaluation of the New Commit: `65c65e38`

The commit `65c65e38` (`fix(mtp): avoid cache rebuilds when requests join (#3702)`) was pushed to upstream main. 

**Verdict:** It is a step in the right direction for active batches, but **it leaves critical blind spots that will still cause the issue in real multi-agent workloads.**

#### Why `65c65e38` Is Incomplete:

1. **The Singleton Host Extension Bug is Still Untouched (Line 192–198)**:
   In `patched_extend`:
   ```python
   host_state = getattr(self, "_omlx_mtp_state", None)
   if host_state is not None and _mtp_state_valid_for_batch(self, host_state):
       ...
       _reconcile_mtp_to_standard(self, host_state) # <-- STILL PRESENT!
       _drop_mtp_state(self, "extend-reconciled")
   ```
   In real workloads, Request 1 starts alone as a singleton (`_omlx_mtp_state`). When Request 2 finishes prefill and joins via `self.extend(batch)`, `host_state` is active. Commit `65c65e38` **still unconditionally calls `_reconcile_mtp_to_standard` on the singleton host**, re-prefilling Request 1's 100k prompt!
   *(jundot's test missed this because `gen.insert(prompts[:2])` inserted two requests at once, starting directly in batch mode rather than extending a singleton).*

2. **The Ragged Queue Fallback in Batch Reconciliation**:
   In `_reconcile_mtp_batch_to_standard`:
   ```python
   if _feed_batch_mains_to_standard(gen_batch, batch_state):
       return True
   for idx, uid in enumerate(gen_batch.uids):
       ...
       if not _reconcile_mtp_to_standard(row, state): # <-- Still falls back to full re-prefill!
   ```
   `_feed_batch_mains_to_standard` fails if `any(state.queue for state in states)`. In speculative execution with variable acceptance across rows, non-empty queues are frequent. Any time a new request arrives while a row has unstreamed queue items, it still falls through to full-prompt re-prefill for every row.

3. **Batch Calibration Steps (`BatchPolicy.needs_standard()`)**:
   Every time the active cohort of requests changes, `BatchPolicy` initializes with `standard_warmup = 2` and `needs_standard() = True`, demanding transitions to standard decode and risking reconciliation re-prefills.

---

### 3. Recommendations & Next Steps

1. You can inform @jundot that:
   * Commit `b45fb7e5` reduced per-chunk overhead but still executes the full `0..total` prompt re-prefill in `_reconcile_mtp_to_standard`. In long-context agentic sessions (70k–150k tokens), re-prefilling hundreds of thousands of tokens blocks the inference thread for minutes, dropping effective TPS to <2 tok/s and inflating wired memory to 240+ GB.
   * Commit `65c65e38` helps the multi-request batch case when queues are empty, but `patched_extend` line 198 still unconditionally calls `_reconcile_mtp_to_standard(self, host_state)` when a second request joins a singleton MTP request.
   * `_reconcile_mtp_to_standard` should never re-prefill the entire prompt history for committed tokens. The cache already holds the prompt; only the 1-token frontier needs alignment (like `_feed_next_main_to_standard`).

**Comment by jundot** [comment]
@shironexus-gif I retested the latest main with Qwen3.6-35B-A3B-oQ6-mtp, starting with one active generation and then submitting several 24K-98K coding contexts concurrently. I verified actual MTP batches of 2-5 requests, cancellation during prefill and decoding, replacement requests, and follow-up turns with prefix cache reuse. I also tested ANE both enabled and disabled.

I couldn't reproduce the reported hang or unexpected full cache rebuilds in these scenarios. If you can reproduce it with non-private context, sharing the prompts, relevant settings, and debug logs covering the stall would be a big help.

**Comment by shironexus-gif** [comment]
@jundot Good Morning :), after adding your commits and clearing every cache that i could possibly find + rebuilding the kernels a second time (just to be sure), everything worked out beautifully.

Thank you so much for spending the time with me resolving this issue. Qwen3.6-35b-a3b is still my main workhorse for my subagents so i am super grateful!
